pkg/virt-controller/watch/replicaset.go

```diff
@@ -184,18 +185,38 @@ func (c *VMReplicaSet) execute(key string) error {
 		return err
 	}
 
-	// make sure we only consider active VMs
 	vms = c.filterActiveVMs(vms)
 
-	// make sure the selector of the controller matches and the VMs match
-	vms = c.filterMatchingVMs(selector, vms)
+	// If any adoptions are attempted, we should first recheck for deletion with
+	// an uncached quorum read sometime after listing VirtualMachines (see kubernetes/kubernetes#42639).
+	canAdoptFunc := controller.RecheckDeletionTimestamp(func() (v1.Object, error) {
+		fresh, err := c.clientset.ReplicaSet(rs.ObjectMeta.Namespace).Get(rs.ObjectMeta.Name, v1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+		if fresh.ObjectMeta.UID != rs.ObjectMeta.UID {
+			return nil, fmt.Errorf("original ReplicaSet %v/%v is gone: got uid %v, wanted %v", rs.Namespace, rs.Name, fresh.UID, rs.UID)
+		}
+		return fresh, nil
+	})
+	cm := controller.NewVirtualMachineControllerRefManager(controller.RealVirtualMachineControl{Clientset: c.clientset}, rs, selector, virtv1.VMReplicaSetGroupVersionKind, canAdoptFunc)
+	vms, err = cm.ClaimVirtualMachines(vms)
+	if err != nil {
+		return err
+	}
 
-	// Scale up or down, if all expected creates and deletes were report by the listener
 	var scaleErr error
-	if needsSync && !rs.Spec.Paused {
+
+	// Scale up or down, if all expected creates and deletes were report by the listener
+	if needsSync && !rs.Spec.Paused && rs.ObjectMeta.DeletionTimestamp == nil {
 		scaleErr = c.scale(rs, vms)
 	}
 
+	// If the controller is going to be deleted and the orphan finalizer is the next one, release the VMs. Don't update the status
+	if rs.ObjectMeta.DeletionTimestamp != nil && len(rs.ObjectMeta.Finalizers) > 0 && rs.ObjectMeta.Finalizers[0] == v1.FinalizerOrphanDependents {
```

Comment:
We shouldn't be making any assumptions that a finalizer is at a specific index.

Reply:
Seems you are right. Double-checked it in the code of kubernetes. It does not look like there is a rule regarding to the finalizer execution order. Surprised me. Will update it.